examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,339 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from copy import deepcopy
+
+import numpy as np
+import torch
+
+import wandb
+from models.actor_critic import init_actor_critic
+from models.reward import init_reward_model
+
+from omegaconf import OmegaConf
+from torch.optim.lr_scheduler import CosineAnnealingLR
+
+from torchrl.data import LazyTensorStorage
+from torchrl.data.replay_buffers import (
+    SamplerWithoutReplacement,
+    TensorDictReplayBuffer,
+)
+from torchrl.data.rlhf.dataset import get_dataloader
+from torchrl.data.rlhf.prompt import PromptData
+from torchrl.data.rlhf.utils import RolloutFromModel
+
+from torchrl.objectives import ClipPPOLoss
+from torchrl.objectives.value import GAE
+from tqdm import tqdm
+from transformers import GenerationConfig, GPT2Tokenizer
+from utils import get_file_logger, resolve_name_or_path, setup
+
+
+def flatten_td(td):
+    # our tensordict has shape [B, T] where B = batch_size and T = trajectory length
+    # some trajectories may have stopped (reached EOS) before generating T tokens
+    # this function truncates and concatenates the trajectories, resulting in a
+    # tensordict that has shape [N] where N <= B * T.
+    done = td["next", "done"]
+    mask = torch.zeros_like(done)
+    mask[..., 1:, :] = done[..., :-1, :]  # shift by one
+    mask = ~mask.cumsum(-2).bool().squeeze()
+    return td[mask]
+
+
+class AdaptiveKLController:
+    """Adaptive KL Controller as described in Ziegler et al. "Fine-Tuning Language Models from Human Preferences"
+    Reference: Section 2.2 https://arxiv.org/pdf/1909.08593.pdf#page=2
+    Source: https://github.com/openai/lm-human-preferences/blob/master/lm_human_preferences/train_policy.py
+    """
+
+    def __init__(self, init_kl_coef: float, target: float, horizon: int):
+        self.value = init_kl_coef
+        self.target = target
+        self.horizon = horizon
+
+    def update(self, current: float, n_steps: int):
+        """Returns adaptively updated KL coefficient, βₜ₊₁.
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
+        """
+        proportional_error = np.clip(current / self.target - 1, -0.2, 0.2)  # ϵₜ
```

Comment:
why np? We should only work with tensors no?

Reply:
it's just a scalar, no?